#1020: Input field type toggles

BLoe opened the pull request:
This PR implements "toggle" inputs on input fields in block configs in the page editor so that a user can choose which type of data is being entered, either a raw literal of some data type, a "wired-up" data variable reference, or a template string.

![image](https://user-images.githubusercontent.com/2801308/140405722-0478e3bf-fe38-4487-ac16-de5b0a6000bc.png)

![image](https://user-images.githubusercontent.com/2801308/140405842-3cca8107-2b56-4a3a-bc44-d624511c3344.png)

twschiller commented on src/components/fields/fieldUtils.ts:
```diff
@@ -48,3 +50,13 @@ export function createTypePredicate(predicate: TypePredicate): TypePredicate {
   };
 }
+
+export function useApiVersion(): ApiVersion {
```
This is very specific to the state of the page editor. It should not go in our generic field utils

twschiller replied:
This method is only used in useApiVersionAtLeast. There's no reason to export it? And it can also just be merged into useApiVersionAtLeast

twschiller commented on src/components/fields/schemaFields/BooleanField.tsx:
```diff
@@ -17,21 +17,20 @@
+  return (
+    <ConnectedFieldTemplate
+      name={name}
+      as={SwitchButtonWidget} // Do not use layout="switch" by default
```
> Do not use layout="switch" by default

Please clarify: why not?

BLoe replied:
I forget where exactly this came from, it's not something new I wrote, I'll remove it, agreed it doesn't make a lot of sense.

twschiller commented on src/components/fields/schemaFields/SchemaField.tsx:
```diff
@@ -15,29 +15,378 @@
+};
+
+function getToggleOptions(
+  fieldSchema: Schema,
```
This signature is very error prone since it includes multiple arguments of the same type (boolean) 

In cases like these it's best to have the function destructure an object so we get named arguments. (And since the names are generally the same in the caller, it can be called with object short hand

twschiller commented on src/components/fields/schemaFields/SchemaField.tsx:
```diff
@@ -15,29 +15,378 @@
+        },
+      },
+      {
+        label: "Handlebars template",
```
As mentioned in our meeting, we can exclude Handlebars since it doesn't work with variable that start with an "@" prefix because it uses that prefix for control flow

twschiller commented on src/components/fields/schemaFields/SchemaField.tsx:
```diff
@@ -15,29 +15,378 @@
+    pushOpts(...anyOfOptions);
+  }
+
+  if (fieldSchema.oneOf?.length > 0) {
```
This looks like it's a duplicate of the one above? Merge them into a single loop 

```
for (const x of [...(fieldSchema.anyOf ?? []), ...(fieldSchema.oneOf ?? [])]) {
  
}
```

twschiller commented on src/components/fields/schemaFields/SchemaField.tsx:
```diff
@@ -15,29 +15,378 @@
+): InputModeOption[] {
+  const options: InputModeOption[] = [];
+
+  function pushOpts(...opts: InputModeOption[]) {
```
NIT: don't abbreviate when the difference is 3-4 letters

twschiller commented on src/components/fields/schemaFields/SchemaField.tsx:
```diff
@@ -15,29 +15,378 @@
+    pushOpts({
+      label: "Remove",
+      value: "omit",
+      symbol: "🗑",
```
As discussed, let's use a difference icon. If you want to use unicode, you could use × or similar

twschiller commented on src/components/fields/schemaFields/SchemaField.tsx:
```diff
@@ -15,29 +15,378 @@
+    []
+  );
+
+  const useLegacyFields = !useApiVersionAtLeast("v3");
```
Negating a hook is wacky

twschiller replied:
I would probably just check this first and split into 2 sub components: 1) LegacySchemaField, and 2) ToggleSchemaField

This will help you avoid some of the wacky checks on useLegacyFields

twschiller commented on src/components/fields/schemaFields/SchemaField.tsx:
```diff
@@ -15,29 +15,378 @@
+      stableSetValue(inputModeOptions[0].defaultValue);
+    }
+  }, [
+    inputModeOptions,
```
Should this only default on initial mount? If so, we can just use a blank `[]` dependency list

twschiller commented on src/components/fields/schemaFields/SelectorSelectorField.tsx:
```diff
@@ -21,13 +21,14 @@ import { fieldLabel } from "@/components/fields/fieldUtils";
+  props
+) => {
+  const { label, name, schema, noLabel } = props;
+  const displayLabel = noLabel ? undefined : label ?? fieldLabel(name);
```
Mixing ternary and nullish coalescing operator is really hard to read (which takes precedence). Add parens or split into 2 statements if prettier is removing the parens

twschiller commented on src/components/fields/schemaFields/fieldInputMode.ts:
```diff
@@ -0,0 +1,68 @@
+  fieldConfig: UnknownObject,
+  fieldName: string
+): FieldInputMode {
+  const hasField = fieldName in fieldConfig;
```
Technically this should be Object.prototype.hasOwnProperty.call(fieldConfig, fieldName)

twschiller commented on src/components/fields/schemaFields/fieldInputMode.ts:
```diff
@@ -0,0 +1,68 @@
+    return value.__type__;
+  }
+
+  if (Array.isArray(value)) {
```
Would be helpful to add a comment here that isArray check has to come before the typeof checks because arrays have type object

twschiller commented on src/components/fields/schemaFields/propTypes.ts:
```diff
@@ -56,4 +61,19 @@ export interface SchemaFieldProps<TValue> {
+  /**
+   * Hide the label
+   */
+  noLabel?: boolean;
```
Why not call hideLabel?

twschiller commented on src/components/fields/schemaFields/propTypes.ts:
```diff
@@ -56,4 +61,19 @@ export interface SchemaFieldProps<TValue> {
+  noLabel?: boolean;
+
+  /**
+   * Is this field a nested object property?
```
Would be helpful to provide context for why you're tracking this

twschiller commented on src/components/fields/schemaFields/propTypes.ts:
```diff
@@ -56,4 +61,19 @@ export interface SchemaFieldProps<TValue> {
+  isObjectProperty?: boolean;
+
+  /**
+   * Is this field an item in an array?
```
Would be helpful to provide context for why you're tracking this in the object.

twschiller commented on src/components/fields/schemaFields/schemaFieldUtils.ts:
```diff
@@ -0,0 +1,32 @@
+    schema: { title },
+    noLabel,
+  } = props;
+  return noLabel ? undefined : label ?? title ?? fieldLabel(name);
```
Please don't mix ternary and null coalescing operator. I think prettier removes the parens because technically the precendence is well-defined. But it's extremely hard to read

twschiller commented on src/components/fields/schemaFields/schemaFieldUtils.ts:
```diff
@@ -0,0 +1,32 @@
+export function makeLabelForSchemaField(
+  props: SchemaFieldProps
+): React.ReactNode {
+  const {
```
Why not destructure in the method signature?

twschiller commented on src/components/fields/schemaFields/widgets/LoadingWidget.tsx:
```diff
@@ -0,0 +1,22 @@
+
+const LoadingWidget: React.FC = () => <div>Loading...</div>;
+
+export default LoadingWidget;
```
Is there any value to calling this a widget vs. WidgetLoadingIndicator or something a bit more descriptive?

twschiller commented on src/components/fields/schemaFields/widgets/NumberWidget.tsx:
```diff
@@ -0,0 +1,23 @@
+import React from "react";
+import { Form } from "react-bootstrap";
+
+const NumberWidget: React.FC = () => <Form.Control type="number" />;
```
Would be nice to add comment referencing the IntegerWidget to help people navigate related fields

```
/**
* @see IntegerWidget
*/
```

twschiller commented on src/components/fields/schemaFields/widgets/TemplateToggleWidget.module.scss:
```diff
@@ -0,0 +1,48 @@
+}
+
+.field {
+  width: auto !important;
```
I'm surprised we need to use important in SCSS modules?

BLoe replied:
The bootstrap styles always override my modules unless I use `!important`

twschiller replied:
@BALEHOK could you chime in here on how to use specificity/global etc to avoid using important? Or is important the right way to go here?

twschiller commented on src/components/fields/schemaFields/widgets/TemplateToggleWidget.tsx:
```diff
@@ -0,0 +1,195 @@
+  ...props
+}) => {
+  const [{ value }, , { setValue }] = useField<unknown>(name);
+  const fieldName = name.includes(".")
```
Refactor into a helper method that returns tuple of the parentFieldName and fieldName

twschiller commented on src/components/fields/schemaFields/widgets/TemplateToggleWidget.tsx:
```diff
@@ -0,0 +1,195 @@
+    ? name.slice(0, name.lastIndexOf("."))
+    : undefined;
+  const { values, setValues } = useFormikContext<UnknownObject>();
+  const parentValues = getIn(values, parentFieldName) ?? values;
```
Why does this default to values?

BLoe replied:
Because `parentFieldName` may be `undefined`, in which case there is no nesting in the field structure, so we should look directly in `values` for the property `name`.

twschiller commented on src/components/fields/schemaFields/widgets/TemplateToggleWidget.tsx:
```diff
@@ -0,0 +1,195 @@
+            }
+          } else if (fieldName in values) {
+            // eslint-disable-next-line @typescript-eslint/no-dynamic-delete,security/detect-object-injection
+            delete values[fieldName];
```
- [ ] This looks like a bug. We don't want to modify the values object. We only want to modify the draft here

BLoe replied:
Nice catch!

twschiller commented on src/components/fields/schemaFields/widgets/TemplateToggleWidget.tsx:
```diff
@@ -0,0 +1,195 @@
+          } else if (fieldName in values) {
+            // eslint-disable-next-line @typescript-eslint/no-dynamic-delete,security/detect-object-injection
+            delete values[fieldName];
+          }
```
Why no else here? I like to include a else which includes a comment as to why the else branch is a no-op

twschiller commented on src/devTools/editor/extensionPoints/base.ts:
```diff
@@ -63,7 +63,7 @@ export interface WizardStep {
  */
 // Keep at v2 until the Page Editor supports v3 expressions
-export const PAGE_EDITOR_DEFAULT_BRICK_API_VERSION: ApiVersion = "v2";
+export const PAGE_EDITOR_DEFAULT_BRICK_API_VERSION: ApiVersion = "v3";
```
I thought we were keeping this at v2 but allowing PixieBrix team members to opt into v3 via toggle in the interface

- [ ] Make page editor default controllable

BLoe replied:
Yep, good catch.

twschiller commented on src/hooks/useBlockOptions.ts:
```diff
@@ -58,9 +57,7 @@ export function useBlockOptions(
-      return (
-        registered ?? genericOptionsFactory(inputProperties(block.inputSchema))
-      );
+      return registered ?? genericOptionsFactory(block.inputSchema);
```
Why isn't inputProperties needed here? Isn't it needed to support the JS bricks using schema shorthand (where they declare properties directly instead of under a properties field?)

BLoe replied:
It's already called inside `genericOptionsFactory()`

twschiller commented on src/components/fields/schemaFields/widgets/TemplateToggleWidget.tsx:
```diff
@@ -0,0 +1,195 @@
+  const parentFieldName = name.includes(".")
+    ? name.slice(0, name.lastIndexOf("."))
+    : undefined;
+  const { values, setValues } = useFormikContext<UnknownObject>();
```
Rename setValues (e.g., setFormValues). It's too similar to setValue.

twschiller commented on src/components/fields/schemaFields/widgets/TemplateToggleWidget.tsx:
```diff
@@ -0,0 +1,195 @@
+      );
+
+      if (newInputMode === "omit") {
+        const newFormState = produce(values, (draft) => {
```
This "delete" field would be a good helper helper method

twschiller reviewed (changes requested):
- See comment about default page editor version
- See comment about potential bug in delete field logic
- Fix broken tests

I think those are the only 2 functional comments. Other comments are about places where documentation/refactoring/renaming would be helpful to make it easier to read and avoid errors when working with the code int he future

twschiller commented on src/components/fields/schemaFields/widgets/ObjectWidget.tsx:
```diff
@@ -190,12 +193,16 @@ export function getPropertyRow(
 
+  // Don't show action on v3 or above
+  const showAction =
+    !useApiVersionAtLeast("v3") && Boolean(additionalProperties);
```
useApiVersionAtLeast is a React hook. Let's not use as if it were it were a normal function call. E.g., if the order was swapped here React would report an error about inconsistent hook execution orders

Assign hook output to a variable and then do any boolean logic

twschiller commented:
@BLoe I added some test cases as partial example of how SchemaField can be tested (at least the static rendering of it, I can also help with test the transitions between types too)

My main concern for getting this PR merged in is not having it impact v1/v2 extensions. Right now that concern is centered on getToggleOptions since the method still runs on v1/v2 extensions. IIRC, I've only see errors when toggling between entry types. But that also worries me for initial rendering as well

See comment https://github.com/pixiebrix/pixiebrix-extension/pull/1768/files#r743337368 about splitting up SchemaField in a way that doesn't execute the v3 code on v1/v2 extensions

BLoe commented:
@twschiller Ready for another pass. I split it into a separate v1 and v3 SchemaField component. I also split the v3 one further into a wrapper that either renders `ServiceField` or `BasicSchemaField` to eliminate another awkward boolean condition in the `useEffect()` call. Let me know if you like that or not, and further, should I keep splitting in this matter for `UnsupportedSchemaField` / `ToggleSchemaField` to eliminate the last awkward boolean condition, or leave it as is?

twschiller commented on src/devTools/editor/hooks/useOmitFormField.tsx:
```diff
@@ -0,0 +1,92 @@
+    // eslint-disable-next-line security/detect-object-injection,@typescript-eslint/no-dynamic-delete
+    delete parent[fieldName];
+  }
+  // Can't remove a field from something that isn't an array or object
```
In what case would a caller call this method if parent is not an array or object?

You should test this path

BLoe replied:
It would be an improper use of this hook, or a mis-configured form state for that to happen. I can log a warning.

twschiller commented on src/devTools/editor/hooks/useOmitFormField.tsx:
```diff
@@ -0,0 +1,92 @@
+  return [parentFieldName, fieldName];
+}
+
+export function removeField(parent: unknown, fieldName: string) {
```
Write return types on exported functions

twschiller commented on src/devTools/editor/hooks/useOmitFormField.tsx:
```diff
@@ -0,0 +1,92 @@
+  onOmitField: () => void;
+};
+
+export function getFieldNamesFromPathString(name: string) {
```
Write return types on exported functions

twschiller commented on src/devTools/editor/hooks/useOmitFormField.tsx:
```diff
@@ -0,0 +1,92 @@
+      } else if (fieldName in formState) {
+        removeField(draft, fieldName);
+      } else {
+        // Cannot find property to remove
```
Why is doing nothing the right thing to do here? Is this case expected to occur in practice? If not we should at least be logging a warning here

twschiller commented on src/devTools/editor/hooks/useOmitFormField.test.tsx:
```diff
@@ -0,0 +1,58 @@
+});
+
+describe("removeField", () => {
+  test("remove field from object", () => {
```
Add test case for non array/object

twschiller reviewed (approved):
Good to go once you address comments on error/unexpected case handling in omit field
